python-squid/src/main/java/org/sonar/python/api/tree/PyDictOrSetLiteralTree.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.api.tree;
+
+import com.sonar.sslr.api.Token;
+import java.util.List;
+
+public interface PyDictOrSetLiteralTree extends PyExpressionTree {
```

Comment:
Given this interface is just a way to factorize code (and btw, factorization by inheritance is not necessarily a good pattern) I believe you could push things one step further with a type parameter : 

```
class PyDictOrSetLiteralTree<T ? extends ExpressionTree>...  {
  List<T> elements();
}
```
and specialize this in the inheritance.

Reply:
Not really sure this is worth it though... I would eventually be more in favor of not complexifying the type hierarchy and I would strongly suggest to avoid this intermediate level in the API as it is not meaningful in the interfaces (those don't share a common thing in term of language construction). (fine in the implementation though, even if it is factorization through inheritance).